### DISCUSSION TITLE
Background process taking too long.

### DISCUSSION THREAD

**Discussion by dega5** [zauberzeug/nicegui#1936] [Q&A]
### Question

The issue is once the user uploads the file and clicks the submit button nothing happens for 10-12 seconds. Then suddenly they are redirected to the task list page. This isn't the best user experience as they don't even know if the button has been clicked or the background process is actually happening. 

Instead I am trying to redirect the user to the 'task-list' page first and then run the background process, but it still runs validate input function before redirecting user to the other page (code below).

```py
ui.upload(
    label='Upload Data Model',
    auto_upload=True,
    multiple=False,
    max_files=1,
    max_file_size=10_000_000,
    on_rejected=lambda: ui.notify('File must a zip, tar or gzip file smaller than 10Mb !'),
    on_upload=on_update_button_click
).props('accept=".gz,.zip,.tar"') 
```

Then I am running this code

```py
def on_update_button_click():
      ui.open('/task-list')
     validate_aci_input()

def validate_aci_input():
    validate_input(region.value, name.value, proj.value)
```

**Comment by rodja** [comment]
I'm not entirely sure about what's the problem and what it is you want to archive. But maybe https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates helps?

**Reply by dega5** [reply]
Sorry if my question wasn't clear. 

The issue is as soon as user clicks the submit button after uploading a file, nothing happens for 10-12 seconds (no loading/ no sign of background process working) and then suddenly, they are redirected to the the task list page. To avoid this bad user experience, I want the user to be redirected to the task-list page first and then run the background process.

  ```py
  def on_update_button_click():
                    ui.open('/task-list')
                   validate_aci_input()


                def validate_aci_input():
                    validate_input(region.value, name.value, proj.value) 
```

Here I have the ui.open('/task-list') command before validate_aci_input(), but it doesnt run in that order. It still runs validate_aci_input() first then ui.open('/task-list'). How can I fix this?

Thanks for your time and help!!

**Reply by rodja** [reply]
> after uploading a file, nothing happens for 10-12 seconds (no loading/ no sign of background process working

Yes. I very much suspect that your `validate_aci_input` is doing cpu-bound or io-bound stuff on the main thread and hence blocks the whole server from processing any other calls. Thats why I provided you with the link to our FAQs. I bet that the error is gone if you replace `validate_aci_input` with `print('validate_aci_input')`.

**Reply by dega5** [reply]
Thanks for your reply!

I changed the code so that ui.open command works first, but now validate_aci_input() function doesn't run at all.
```
              async def on_update_button_click():
                    await ui.open('/task-list')
                    validate_aci_input()


                def validate_aci_input():
                    validate_input(region.value, name.value, proj.value)
                    
```

**Reply by rodja** [reply]
You need to make sure nothing "demanding" is executed on the main thread. What are you doing in `validate_aci_input`? Try using `await run.cpu_bound(validate_aci_input)`.

**Reply by dega5** [reply]
The function calls a very demanding background process so it makes sense for it to take 10/12 seconds.
I get an error mesage - 'nicegui.run has no attribute cpi_bound'

**Reply by rodja** [reply]
Please read https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates carefully and follow the links.

**Reply by dega5** [reply]
I have tried to follow both these links, and tried the different options but I am still struggling. I am new to programming thats why I dont really understand the deep technical stuff. 


[https://fastapi.tiangolo.com/async/#in-a-hurry ](https://fastapi.tiangolo.com/async/#in-a-hurry)
[https://nicegui.io/documentation#running_cpu-bound_tasks ](https://nicegui.io/documentation#running_cpu-bound_tasks)

**Reply by natankeddem** [reply]
Does your processing routine consist of iterating over a dataset? If so you could yield periodically during this iteration inside an async function.
https://stackoverflow.com/questions/36647825/cooperative-yield-in-asyncio

**Comment by natankeddem** [comment]
How about using `ui.spinner`?
```py
from nicegui import ui, events
import asyncio

async def on_update_button_click(e: events.UploadEventArguments):
    e.sender.visible = False
    await validate_aci_input()
    ui.open("/task-list")

async def validate_aci_input():
    await asyncio.sleep(5)

@ui.page("/")
def root():
    upload = ui.upload(
        label="Upload Data Model",
        auto_upload=True,
        multiple=False,
        max_files=1,
        max_file_size=10_000_000,
        on_rejected=lambda: ui.notify("File must a zip, tar or gzip file smaller than 10Mb !"),
        on_upload=lambda e: on_update_button_click(e),
    ).props('accept=".gz,.zip,.tar"')
    ui.spinner(size="xl").bind_visibility_from(upload, "visible", lambda x: not x)

@ui.page("/task-list")
def task_list():
    ui.label("task-list")

ui.run()
```

**Reply by dega5** [reply]
This doesn't work unfortunately - I don't see the spinner at all.
Its the same as before, once I click the submit button nothing happens for 10/12 seconds and then suddenly I am redirected to the task list page.

**Reply by natankeddem** [reply]
Are you running exactly what I posted above? This is how it looks on my computer for reference:
[google_screen_recording_2023-11-02T17-04_03.698Z.webm](https://github.com/zauberzeug/nicegui/assets/44515217/0c6ebda6-b55b-4cdf-8969-a93b07bdf51c)

**Reply by dega5** [reply]
I am not running it exactly like that because my background process is different and won't be able to share too much because of data security.

**Reply by natankeddem** [reply]
My point is this, if you run my code unaltered you should see a spinner. If you don't there might be something wrong with your environment, since it runs on my computer correctly (see post above for video evidence). Now you have established a working baseline. After that we can talk about alterations of the code to accommodate your processing routine.

**Comment by rodja** [comment]
This might be solvable with #2410.
